proto/pulumi/provider.proto

```diff
@@ -594,35 +603,78 @@ message DiffResponse {
+    string name = 5;
+
+    // The type of the resource being created. This must match the type specified by the `urn` field, and is passed so
+    // that providers do not have to implement URN parsing in order to extract the type of the resource.
+    string type = 6;
 }
 
+// `CreateResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.Create) call. A `CreateResponse`
+// contains the ID of the created resource, as well as any output properties that arose from the creation process.
 message CreateResponse {
-    // NOTE: The partial-update-error equivalent of this message is `ErrorResourceInitFailed`.
+    // The ID of the created resource.
+    string id = 1;
 
-    string id = 1;                         // the ID of the created resource.
-    google.protobuf.Struct properties = 2; // any properties that were computed during creation.
+    // The resource's output properties. Typically this will be a union of the resource's input properties and any
+    // additional values that were computed or made available during creation.
+    google.protobuf.Struct properties = 2;
 }
 
+// `ReadRequest` is the type of requests sent as part of a [](pulumirpc.ResourceProvider.Read) call.
 message ReadRequest {
-    string id = 1;                         // the ID of the resource to read.
-    string urn = 2;                        // the Pulumi URN for this resource.
-    google.protobuf.Struct properties = 3; // the current state (sufficiently complete to identify the resource).
-    google.protobuf.Struct inputs = 4;     // the current inputs, if any (only populated during refresh).
-    string name = 5; // the Pulumi name for this resource.
-    string type = 6; // the Pulumi type for this resource.
+    // The ID of the resource to read.
+    string id = 1;
+
+    // The URN of the resource being read.
+    string urn = 2;
+
+    // Any current state for the resource being read. This state should be sufficient to uniquely identify the resource.
+    google.protobuf.Struct properties = 3;
+
+    // Any current input properties for the resource being read. These will only be populated when the
+    // [](pulumirpc.ResourceProvider.Read) call is being made as part of a refresh operation.
+    google.protobuf.Struct inputs = 4;
+
+    // The name of the resource being read. This must match the name specified by the `urn` field, and is passed so that
+    // providers do not have to implement URN parsing in order to extract the name of the resource.
+    string name = 5;
+
+    // The type of the resource being read. This must match the type specified by the `urn` field, and is passed so that
+    // providers do not have to implement URN parsing in order to extract the type of the resource.
+    string type = 6;
 }
 
+// `ReadResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.Read) call. A `ReadResponse` contains
```

Comment:
We must document how to indicate that a resource could not be found in the cloud. It is super common.

Reply:
Yep I've added a general "what are the error cases?" point to #17691. I have a lot of digging to do xD